linz2osm/convert/osm.py

```diff
@@ -378,33 +432,72 @@ def export_custom(layer_in_dataset, feature_ids = None, workslice_id = None):
                        special_start_node_field_name=layer.special_start_node_field_name,
                        special_end_node_field_name=layer.special_end_node_field_name
                        )
+
+    member_feature_map = {}
+
+    # None of this happens unless the layer is a RELATION
+    for m in layer.members.all():
+        m_layer = m.member_layer
+        m_processors = m_layer.get_processors()
+        for i, (row_data, row_geom) in enumerate(member_data_tables[m.pk]):
+            db_feature_id = row_data[m_layer.pkey_name]
+            db_table_name = m_layer.name
+            if m_layer.geometry_type == 'POINT':
+                osm_feature_type = 'node'
+            elif m_layer.geometry_type == 'LINESTRING':
+                osm_feature_type = 'way'
+            else:
+                raise NotImplementedError("We do not yet support %ss as members" % m_layer.geometry_type)
+            key = (db_table_name, db_feature_id)
+            # Reuse a member feature if already included as a different role.
+            if key not in member_feature_map:
+                osm_feature_ids = _export_custom_data_row(writer, m_layer, member_tags[m.pk], m_processors, i, row_data, row_geom)
+                member_feature_map[key] = (osm_feature_type, osm_feature_ids)
+
     for i, (row_data, row_geom) in enumerate(data_table):
-        row_tags = []
-        for tag in layer_tags:
-            try:
-                v = tag.eval(row_data)
-            except tag.ScriptError, e:
-                emsg = "Error evaluating '%s' tag against record:\n" % tag
-                emsg += json.dumps(e.data, indent=2) + "\n"
-                emsg += str(e)
-                raise ValueError(emsg)
-            if (v is not None) and (v != ""):
-                row_tags.append((tag.tag, v, tag,))
-
-        # apply geometry processing
-        for p in processors:
-            row_geom = p.process(row_geom, fields=row_data, tags=row_tags, id=i)
-
-        if layer.special_node_reuse_logic:
-            first_node_ref = str(row_data.get(layer.special_start_node_field_name))
-            last_node_ref = str(row_data.get(layer.special_end_node_field_name))
-        else:
-            first_node_ref, last_node_ref = None, None
+        member_refs = []
+        for m in layer.members.all():
```

Comment:
Worth doing a layer.members.all().select_related("member_layer") to avoid 2N+1 DB queries here?